src/hugr/rewrite/simple_replace.rs

```diff
@@ -41,6 +48,198 @@ impl SimpleReplacement {
     }
 }
 
+impl Rewrite<SimpleReplacementError> for SimpleReplacement {
+    const UNCHANGED_ON_FAILURE: bool = true;
+
+    fn verify(&self, _h: &Hugr) -> Result<(), SimpleReplacementError> {
+        unimplemented!()
+    }
+
+    fn apply(self, h: &mut Hugr) -> Result<(), SimpleReplacementError> {
+        // 1. Check the parent node exists and is a DFG node.
+        if h.get_optype(self.parent).tag() != OpTag::Dfg {
+            return Err(SimpleReplacementError::InvalidParentNode());
+        }
+        // 2. Check that all the to-be-removed nodes are children of it and are leaves.
+        for node in &self.removal {
+            if h.hierarchy.parent(node.index) != Some(self.parent.index)
+                || h.hierarchy.has_children(node.index)
+            {
+                return Err(SimpleReplacementError::InvalidRemovedNode());
+            }
+        }
+        // 3. Do the replacement.
+        // 3.1. Add copies of all replacement nodes and edges to h. Exclude Input/Output nodes.
+        // Create map from old NodeIndex (in self.replacement) to new NodeIndex (in self).
+        let mut index_map: HashMap<NodeIndex, NodeIndex> = HashMap::new();
+        let replacement_nodes = self
+            .replacement
+            .children(self.replacement.root())
+            .collect::<Vec<Node>>();
+        // number of replacement nodes including Input and Output:
+        let replacement_sz = replacement_nodes.len();
+        // slice of nodes omitting Input and Output:
+        let replacement_inner_nodes = &replacement_nodes[1..replacement_sz - 1];
+        for &node in replacement_inner_nodes {
+            // Check there are no const inputs.
+            if !self
+                .replacement
+                .get_optype(node)
+                .signature()
+                .const_input
+                .is_empty()
+            {
+                return Err(SimpleReplacementError::InvalidReplacementNode());
+            }
+        }
+        let self_input_node_index = h.hierarchy.first(self.parent.index).unwrap();
+        let replacement_output_node = *replacement_nodes.last().unwrap();
+        for &node in replacement_inner_nodes {
+            // Add the nodes.
+            let op: &OpType = self.replacement.get_optype(node);
+            let sig = op.signature();
+            let new_node_index = h.graph.add_node(sig.input.len(), sig.output.len());
+            h.op_types[new_node_index] = op.clone();
+            // Make self.parent the parent
+            h.hierarchy
+                .insert_after(new_node_index, self_input_node_index)
+                .ok();
```

Comment:
@cqc-alec all this code was basically moved here, modulo renaming `self` -> `h` and `r` -> `self`. I'm not sure about this `.ok()` though - that converts the Result to an Option, and then throws away. Should we '.unwrap'? Happy to make a driveby change in this PR if you agree

Reply:
Yes happy with the drive-by change.